Block UnmanagedCallersOnly in VB

333fred opened the pull request:
Fixes https://github.com/dotnet/roslyn/issues/47719. @dotnet/roslyn-compiler @AlekseyTs for review.

@wli3 this has a string change for VB. Is it too late for 16.9 for this? This is an extremely rare error that I don't realistically expect most people to hit.

AlekseyTs commented:
>             ' Flags are packed into a 32-bit int with the following layout:

Please adjust the layout comment.

---
Refers to: src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEMethodSymbol.vb:51 in caefe13. [](commit_id = caefe13b67dc01277b9ef3476f5a0a9fa012bfbd, deletion_comment = False)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEMethodSymbol.vb:
```diff
@@ -178,6 +192,12 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
                 ThreadSafeFlagOperations.Set(_bits, bitsToSet)
             End Sub
+
+            Public Sub SetIsUnmanagedCallersOnly(isUnmanagedCallersOnly As Boolean)
```
>SetIsUnmanagedCallersOnly [](start = 23, length = 25)

"InitializeIsUnmanagedCallersOnly"? Consider following naming convention for other methods changing a pair of related bits.

333fred replied:
No longer initializing a pair of related bits.

---
In reply to: [566915186](https://github.com/dotnet/roslyn/pull/50886#discussion_r566915186) [](ancestors = 566915186)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEMethodSymbol.vb:
```diff
@@ -1136,12 +1156,28 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+        Private Sub CheckUnmanagedCallersOnly(ByRef errorInfo As DiagnosticInfo)
+            If Not _packedFlags.IsUnmanagedCallersOnlyPopulated Then
+                Dim hasUnmanagedCallersOnly As Boolean =
+                    DirectCast(ContainingModule, PEModuleSymbol).Module.FindLastTargetAttribute(_handle, AttributeDescription.UnmanagedCallersOnlyAttribute).HasValue
```
>FindLastTargetAttribute [](start = 72, length = 23)

Do we actually need to find the last application or any would be sufficient for our purpose?

333fred replied:
> Do we actually need to find the last application or any would be sufficient for our purpose?

Any would be sufficient. I'll make the find first target attribute helper internal, it's private today.

---
In reply to: [566918784](https://github.com/dotnet/roslyn/pull/50886#discussion_r566918784) [](ancestors = 566918784)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEMethodSymbol.vb:
```diff
@@ -1136,12 +1156,28 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+                _packedFlags.SetIsUnmanagedCallersOnly(hasUnmanagedCallersOnly)
+            End If
+
+            If _packedFlags.IsUnmanagedCallersOnly Then
```
>If _packedFlags.IsUnmanagedCallersOnly Then [](start = 12, length = 43)

It looks like this is the only place where use this flag and it actually "behind" ```IsUseSiteDiagnosticPopulated``` flag. Therefore, I don't thing we need those two new bits in ```_packedFlags```.

AlekseyTs replied:
> It looks like this is the only place where use this flag and it actually "behind" IsUseSiteDiagnosticPopulated flag. Therefore, I don't think we need those two new bits in _packedFlags.

This doesn't look addressed. We do not need to add any bits in my opinion. So both of them (two) should be removed.

---
In reply to: [566921884](https://github.com/dotnet/roslyn/pull/50886#discussion_r566921884) [](ancestors = 566921884)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEMethodSymbol.vb:
```diff
@@ -1136,12 +1156,28 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+
+            If _packedFlags.IsUnmanagedCallersOnly Then
+                If errorInfo Is Nothing OrElse errorInfo.Code <> ERRID.ERR_UnsupportedMethod1 Then
+                    errorInfo =
```
>errorInfo = [](start = 20, length = 11)

It feels like the attribute check can be moved inside this ```If```.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Emit/Attributes/AttributeTests_UnmanagedCallersOnly.vb:
```diff
@@ -0,0 +1,404 @@
+Imports Roslyn.Test.Utilities
+Imports Roslyn.Test.Utilities.TestMetadata
+
+Namespace Microsoft.CodeAnalysis.VisualBasic.UnitTests.Semantics
```
>Semantics [](start = 55, length = 9)

It looks like we are in Emit unit-tests.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Emit/Attributes/AttributeTests_UnmanagedCallersOnly.vb:
```diff
@@ -0,0 +1,404 @@
+Imports Roslyn.Test.Utilities.TestMetadata
+
+Namespace Microsoft.CodeAnalysis.VisualBasic.UnitTests.Semantics
+    Public Class AttributeTests_UnmanagedCallersOnly
```
I am not sure if Emit unit-tests is the right place for these tests. Are they actually validating emit behavior? Perhaps the file should be moved to Symbols tests?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/VBResources.resx:
```diff
@@ -5626,4 +5626,7 @@
     <value>Init-only '{0}' cannot be implemented.</value>
   </data>
+  <data name="ERR_UnmanagedCallersOnlyNotSupported" xml:space="preserve">
+    <value>'UnmanagedCallersOnly' is not supported.</value>
```
>'UnmanagedCallersOnly' is not supported. [](start = 11, length = 40)

"'UnmanagedCallersOnly' attribute is not supported."?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Emit/Attributes/AttributeTests_UnmanagedCallersOnly.vb:
```diff
@@ -0,0 +1,404 @@
+
+
+        'Dim c As New C()
+        'c.Extension()
```
Are these lines needed?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Emit/Attributes/AttributeTests_UnmanagedCallersOnly.vb:
```diff
@@ -0,0 +1,404 @@
+public class C
+{
+    [UnmanagedCallersOnly]
+    public static void M1() { }
```
>public static void M1() { } [](start = 4, length = 27)

Consider explicitly asserting that this symbol has a use site error and also has unsupported metadata.

333fred replied:
Did this for every from metadata test as well.

---
In reply to: [566929962](https://github.com/dotnet/roslyn/pull/50886#discussion_r566929962) [](ancestors = 566929962)

AlekseyTs commented:
Done with review pass (commit 1)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [769900731](https://github.com/dotnet/roslyn/pull/50886#issuecomment-769900731) [](ancestors = 769900731)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEMethodSymbol.vb:
```diff
@@ -1136,12 +1150,25 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
+            If Not _packedFlags.IsUseSiteDiagnosticPopulated AndAlso (errorInfo Is Nothing OrElse errorInfo.Code <> ERRID.ERR_UnsupportedMethod1) Then
+                Dim hasUnmanagedCallersOnly As Boolean =
+                    DirectCast(ContainingModule, PEModuleSymbol).Module.FindTargetAttribute(_handle, AttributeDescription.UnmanagedCallersOnlyAttribute).HasValue
+                _packedFlags.SetIsUnmanagedCallersOnly(hasUnmanagedCallersOnly)
```
>SetIsUnmanagedCallersOnly [](start = 29, length = 25)

I don't think we need to do this.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEMethodSymbol.vb:
```diff
@@ -1136,12 +1150,25 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
         End Function
 
+        Private Sub CheckUnmanagedCallersOnly(ByRef errorInfo As DiagnosticInfo)
+            If Not _packedFlags.IsUseSiteDiagnosticPopulated AndAlso (errorInfo Is Nothing OrElse errorInfo.Code <> ERRID.ERR_UnsupportedMethod1) Then
```
>Not _packedFlags.IsUseSiteDiagnosticPopulated AndAlso [](start = 15, length = 53)

I don't think we need to check this. If you going after optimization, I would prefer us to not call this method rather than changing its outcome based on optimization.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PEMethodSymbol.vb:
```diff
@@ -75,6 +76,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
             Private Const s_isConditionalAttributePopulatedBit As Integer = 1 << 11
             Private Const s_isInitOnlyBit = 1 << 12
             Private Const s_isInitOnlyPopulatedBit = 1 << 13
+            Private Const s_isUnmanagedCallersOnlyBit = 1 << 14
```
>Private Const s_isUnmanagedCallersOnlyBit = 1 << 14 [](start = 12, length = 51)

I do not think we need this bit.

AlekseyTs replied:
> I do not think we need this bit.

Marked as resolved, but it doesn't look addressed.

---
In reply to: [567035007](https://github.com/dotnet/roslyn/pull/50886#discussion_r567035007) [](ancestors = 567035007)

AlekseyTs commented:
Done with review pass (commit 3)

AlekseyTs commented:
Done with review pass (commit 4)

AlekseyTs reviewed (approved):
LGTM (commit 4), assuming CI is passing.

cston commented on src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMethodSymbol.vb:
```diff
@@ -1576,6 +1576,9 @@ lReportErrorOnTwoTokens:
             If attrData.IsTargetAttribute(Me, AttributeDescription.TupleElementNamesAttribute) Then
                 arguments.Diagnostics.Add(ERRID.ERR_ExplicitTupleElementNamesAttribute, arguments.AttributeSyntaxOpt.Location)
+            ElseIf attrData.IsTargetAttribute(Me, AttributeDescription.UnmanagedCallersOnlyAttribute) Then
+                ' VB does not support UnmanagedCallersOnly definitions on methods at all
```
Should this be "... attributes" rather than "... definitions"?

333fred commented:
@jaredpar for approval to merge into 16.9.

333fred commented:
Retargeted for 16.10.

craigajohnson commented:
How about we not do this

333fred commented:
VB cannot do pretty much anything involving unsafe code. This is no different.
